src/qibo/optimizers.py

```diff
@@ -109,4 +124,144 @@ def opt_step():
         if e % sgd_options["nmessage"] == 1:
             log.info('ite %d : loss %f', e, l.numpy())
 
-    return loss(vparams).numpy(), vparams.numpy()
+    return loss(vparams, *args).numpy(), vparams.numpy()
+
+
+class ParallelBFGSResources: # pragma: no cover
+    """Auxiliary singleton class for sharing memory objects in a
+    multiprocessing environment when performing a parallel_L-BFGS-B
+    minimization procedure.
+
+    This class takes care of duplicating resources for each process
+    and calling the respective loss function.
+    """
+    import multiprocessing as mp
+
+    # private objects holding the state
+    _instance = None
+    # dict with of shared objects
+    _objects_per_process = {}
+    custom_loss = None
+    lock = None
+    args = ()
+
+    def __new__(cls, *args, **kwargs):
+        """Creates singleton instance."""
+        if cls._instance is None:
+            cls._instance = super(ParallelBFGSResources, cls).__new__(cls, *args, **kwargs)
+        return cls._instance
+
+    def loss(self, params):
+        """Computes loss (custom_loss) for a specific set of parameters.
+        This class performs the lock mechanism to duplicate objects
+        for each process.
+        """
```

Comment:
If I understand correctly, this class is fairly general in the sense that it can be used to parallelize the calculation of any function assigned as the `custom_loss` variable, right? Perhaps it can be generalized slightly more by making the `params` argument optional.

The reason is that this could be useful in repeated circuit evaluations, such as these we implemented for noise. In this case we just need to execute the same circuit (without changing any parameters) several times and collect the results. So in this case we would have something like `custom_loss = lambda c: c()`. We would also need to collect the results from all processes somehow when calling using `self.pool`.